tests/stack/test_api.py

```diff
@@ -0,0 +1,238 @@
+            extension and the filename of the stack configuration file.
+        """
+        config_path = os.path.join(tmpdir.strpath, 'test.json')
+        expected_status_path = os.path.join(tmpdir.strpath, 'test.deployed.json')
+        generated_path = stack_api._generate_stack_status_path(config_path)
+        assert expected_status_path == generated_path
+
+        config_path = os.path.join(tmpdir.strpath, 'test.st-ack.json')
+        expected_status_path = os.path.join(tmpdir.strpath, 'test.st-ack.deployed.json')
+        generated_path = stack_api._generate_stack_status_path(config_path)
+        assert expected_status_path == generated_path
+
+    def test_save_load_stack_status(self, stack_api, tmpdir):
+        """
+            When saving the a stack status through _save_stack_status, it should be able to be
+            loaded by _load_stack_status and have the same exact contents.
+        """
+        config_path = os.path.join(tmpdir.strpath, 'test.json')
+        status_path = stack_api._generate_stack_status_path(config_path)
+        stack_api._save_json(status_path, TEST_STATUS)
+
+        status = stack_api._load_json(status_path)
+        assert status == TEST_STATUS
+
+    def test_deploy_relative_paths(self, stack_api, tmpdir):
+        """
+            When doing stack_api.deploy, in every call to stack_api._deploy_resource, the current
+            working directory should be the same directory as where the stack config template is
+            contained so that relative paths for resources can be relative to the stack config
+            instead of where CLI calls the API functions.
+        """
+        config_working_dir = os.path.join(tmpdir.strpath, 'stack')
+        config_path = os.path.join(config_working_dir, 'test.json')
+        os.makedirs(config_working_dir)
+        with open(config_path, 'w+') as f:
+            json.dump(TEST_STACK, f)
+
+        def _deploy_resource(resource, stack_status):
+            assert os.getcwd() == config_working_dir
+            return TEST_JOB_STATUS
+
+        stack_api._deploy_resource = mock.Mock(wraps=_deploy_resource)
+        stack_api.deploy(config_path)
+
+    def test_deploy_job(self, stack_api):
+        """
+            stack_api._deploy_job should create a new job when 1) A physical_id is not given and
+            a job with the same name does not exist in the settings.
+
+            stack_api._deploy_job should reset/update an existing job when 1) A physical_id is given
+            2) A physical_id is not given but one job with the same name exists.
+
+            A StackError should be raised when 1) A physical_id is not given but there are multiple
+            jobs with the same name that exist.
+        """
+        test_job_settings = TEST_JOB_SETTINGS
+        alt_test_job_settings = {'name': 'alt test job'}  # Different name than TEST_JOB_SETTINGS
+        jobs_in_databricks = {}
```

Comment:
This in-memory state and the mock jobs client should be lifted into fixtures called like `InMemoryJobs` and `InMemoryJobsClient`